### PR TITLE
Created concept entry for SwiftUI Notifications

### PR DESCRIPTION
**Pull request by arirays**
### Description
- Added SwiftUI Notifications concept entry
- Closes issue #2682

### Type of Change

- Adding a new entry
- Updating the documentation

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] Under "Development" on the right, I have linked any issues that are relevant to this PR (write "Closes #<issue number> in the "Description" above).

### CODE REVIEW

File: content/swiftui/concepts/notifications/notifications.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
+---
+Title: 'Notifications'
+Description: 'Notifications in SwiftUI send notifications to the user's device either from a server or create them directly within an app.'
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:3]:
```suggestion
Description: 'It is used to send notifications to the user's device either from a server or create them directly within an app.' 
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+  - 'Mobile Development'
+Tags: 
+  - 'iOS'
+  - 'SwiftUI'
+CatalogContent:
+  - 'learn-ios-course'
+  - 'paths/ios-developer'
+---
+
+**Notifications** in SwiftUI are presented with two options types to keep users informed and engaged: Local notifications and Remote(push) notifications. Local notifications differ from remote (push) notifications in that they don't depend on a server to deliver messages to users' devices. While remote (push) notifications use Apple's Push Notification Service (APNS) for message delivery, local notifications are self-contained within the app.
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:15]:
```suggestion
**Notifications** in SwiftUI are presented with two option types to keep users informed and engaged: **Local notifications** and **Remote(push) notifications**. **Local notifications** differ from remote (push) notifications in that they don't depend on a server to deliver messages to users' devices. While **remote (push) notifications** use Apple's Push Notification Service (APNS) for message delivery, local notifications are self-contained within the app. 
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+Definition: Local notifications are created and scheduled within the app itself, without any need for an external server. They are ideal for reminders, calendar events, or any other notifications that are triggered by the app based on user actions or predefined schedules.
+
+Scheduling: The developer sets up the timing and content of these notifications. They can be scheduled to appear at a specific time or in response to a particular event within the app.
+
+User permissions: Though less intrusive than remote notifications, local notifications still require user permission. The app must request and obtain permission to send notifications to the user.
+
+Customization: Local notifications can be customized with various options such as sound, alerts, or badges. They can also be configured to trigger certain actions when interacted with.
+
+Implementation in SwiftUI: Implementing local notifications in SwiftUI involves using the UNUserNotificationCenter to manage notification content, request permission, and schedule the notification delivery.
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:28]:
```suggestion
- `Definition`: Local notifications are created and scheduled within the app itself, without any need for an external server. They are ideal for reminders, calendar events, or any other notifications that are triggered by the app based on user actions or predefined schedules.

- `Scheduling`: The developer sets up the timing and content of these notifications. They can be scheduled to appear at a specific time or in response to a particular event within the app.

- `User permissions`: Though less intrusive than remote notifications, local notifications still require user permission. The app must request and obtain permission to send notifications to the user.

- `Customization`: Local notifications can be customized with various options such as sound, alerts, or badges. They can also be configured to trigger certain actions when interacted with.

- `Implementation in SwiftUI`: Implementing local notifications in SwiftUI involves using the UNUserNotificationCenter to manage notification content, request permission, and schedule the notification delivery.
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    // 5. Add the request to the notification center
+    UNUserNotificationCenter.current().add(request) { error in
+        if let error = error {
+            print(error.localizedDescription)
+        } else {
+            print("Notification scheduled")
+        }
+    }
+}
+
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:67]:
```suggestion
func scheduleTimeBasedNotification() {
  // 1. Request permission to display alerts and play sounds.
  UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, error in
    if granted {
        print("Permission granted")
    } else if let error = error {
        print(error.localizedDescription)
    }
}

  // 2. Create the content for the notification
  let content = UNMutableNotificationContent()
  content.title = "Reminder"
  content.body = "Don't forget to check the app!"
  content.sound = UNNotificationSound.default

  // 3. Set up a trigger for the notification
  // For example, 10 seconds from now
  let trigger = UNTimeIntervalNotificationTrigger(timeInterval: (10), repeats: false)

  // 4. Create the request
  let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: trigger)

  // 5. Add the request to the notification center
  UNUserNotificationCenter.current().add(request) { error in
    if let error = error {
        print(error.localizedDescription)
    } else {
        print("Notification scheduled")
    }
  }
}
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+Server interaction and APNS: Remote notifications require a backend server that sends the notification payload to Apple’s Push Notification Service (APNS), which then delivers the message to the appropriate device.
+
+User permissions and settings: As with local notifications, user permission is required. The app must explicitly ask the user to allow push notifications. Users can also customize their settings for these notifications.
+
+Content and interaction: Remote notifications can include media like images and sounds and can provide options for user interactions, such as replying to a message directly from the notification.
+
+Implementation in SwiftUI: Implementing push notifications in a SwiftUI app involves configuring the app to receive remote notifications, handling the registration process with APNS, and then dealing with the incoming notification data appropriately.
+
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:81]:
```suggestion
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    
+    var body: some Scene {
+        WindowGroup {
+            ContentView()
+        }
+    }
+}
+```
+
+![SwiftUI Remote Push Notifications](../../../../media/swiftui-remote-push-notification.png)
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:139]:
```suggestion
![SwiftUI Remote Push Notifications](https://raw.githubusercontent.com/Codecademy/docs/main/media/swiftui-remote-push-notification.png)
```

File: content/swiftui/concepts/notifications/notifications.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        let tokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
+        print("Device push notification token: \(tokenString)")
+        // Here you would typically send the token to your server
+    }
+
+    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+        print("Failed to register for remote notifications: \(error)")
+    }
+}
+
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:118]:
```suggestion
class AppDelegate: NSObject, UIApplicationDelegate {

  func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .badge, .sound]) { granted, error in
        if granted {
            DispatchQueue.main.async {
                UIApplication.shared.registerForRemoteNotifications()
            }
        }
    }
    return true
}

func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    // Convert token to string and send to your server
    let tokenString = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
    print("Device push notification token: \(tokenString)")
    // Here you would typically send the token to your server
}

  func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
      print("Failed to register for remote notifications: \(error)")
  }
}
```

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
Have few comments for you to work on @arirays 🚀

File: content/swiftui/concepts/notifications/notifications.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+  UNUserNotificationCenter.current().add(request) { error in
+    if let error = error {
+        print(error.localizedDescription)
+    } else {
+        print("Notification scheduled")
+    }
+  }
+}
+```
+
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:68]: Is there a possibility for you to add an image example for this notification type here?

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
I have one last question for you @arirays

File: content/swiftui/concepts/notifications/notifications.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  UNUserNotificationCenter.current().add(request) { error in
+    if let error = error {
+        print(error.localizedDescription)
+    } else {
+        print("Notification scheduled")
+    }
+  }
+}
+```
+![Simulator Screenshot - iPhone 15 Pro - 2023-12-20 at 16 41 31](https://github.com/arirays/docs/assets/51388551/ded8763a-245f-4ea3-af16-d680b7d0bec4)
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:68]: can you add this image in the media and make this link like this:
https://raw.githubusercontent.com/Codecademy/docs/main/media/background-position-example-1.png

File: content/swiftui/concepts/notifications/notifications.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+  @UIApplicationDelegateAdaptor private var appDelegate: AppDelegate
+  
+  var body: some Scene {
+    WindowGroup {
+        ContentView()
+    }
+  }
+}
+```
+![Simulator Screenshot - iPhone 15 Pro - 2023-12-02 at 11 50 11](https://github.com/arirays/docs/assets/51388551/edad365a-7d4c-4299-945b-c2781a243744)
```
Comment (dakshdeepHERE) [inline content/swiftui/concepts/notifications/notifications.md:136]: same here check the above comment please

**Review by dakshdeepHERE — CHANGES_REQUESTED** [review]
Please check my last comments

**Review by dakshdeepHERE — APPROVED** [review]
Thank you @arirays for working on the comments. This looks good for second review 🚀

**Comment by arirays** [comment]
Hi Steven @SSwiniarski, could you please review my PR?
And I have a question about my previous contributions, I see my entries posted at Codecademy, but as an anonymous contributor. I followed all contributions guidelines, but unfortunately my profile isn't visible. Is it possible to change it?
I'd really appreciate it if you could help me. Thank you!

**Comment by dakshdeepHERE** [comment]
@arirays I'll let @SSwiniarski know. Also, you'll have to make your Codecademy profile public and link it with this Github email. Once linking the github is done with Codecademy account once log it out and then login using Github account. Then you can see your account on the entries. Make sure the mails are same for github.

**Comment by SSwiniarski** [comment]
@dakshdeepHERE any idea on the failing check? I can't seem to get rid of it.

**Comment by dakshdeepHERE** [comment]
@SSwiniarski Done, I think there was some issue with the npm and yarn update. After updating it got fixed for me.

**Pull request merged**